### PR TITLE
feat: verify block and state proof

### PR DESCRIPTION
**Pull request by vuvoth**
# Context 
- Verify consensus state 
- State proof

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,3 +1,19 @@
...
+
+## Cli interface 
+
+### Fetch Consensus State and Header 
+This command fetch consensus state and header from full-node. This only use for testing `Verify` command.  
+
+### Verify 
+
+```bash
+tendermint-lightclient verify <CS_PATH> <HEADER_PATH> <NEW_CS_PATH>
```
Comment (robert-zaremba) [inline README.md]: Could you describe the format of consensus state?
Is it verifying or updating the consensus? Based on the usage I guess it updates, but the section title says it is verifying

File: README.md [hunk 2]
```diff
@@ -1,3 +1,19 @@
...
+```
+
+Verify a new state (can extract from header) is valid state from consensus state in cs_path file. This also create new consesus state and save to new_cs_path. 
+
+### State Proof 
+In current version, we only support proof packet transfer on full-node. We only verify the Cosmos Store.
```
Comment (robert-zaremba) [inline README.md]:
```suggestion

In current version, we only support proof packet transfer on full-node. We only verify the Cosmos IAVL Store.  
```

File: src/context.rs [hunk 3]
```diff
@@ -294,4 +301,74 @@ mod tests {
...
...N5yIwCAESCSCu0wKah7wRIBohIEpnFv1GjFbFsqE429zfZiKfOT6//q45DUFSPLrY9QoDIi4IARIqIpjxBZqHvBEgQaWrGBvLw429u+E3zsCC20seuQ2K4Np8CgTQd8VhLEQgIjAIARIJJKTOB5qHvBEgGiEg/7viCbl3Q1jvaUZEPFz/2gw7knpIfR+pUpQfbqemvM8iLggBEiomuKMPmoe8ESBx0HzWr2/H5w2JBV7mGKRoCT2nY3baDGbXYLiB/hIYoCAiMAgBEgkq6Jwhmoe8ESAaISAgRLXnK55+pWAZPGrk6b1ZPtOwMcyynsQFqaJmY1SkyiIwCAESCSygyUmah7wRIBohIDzvjMPa1Sr0H1VwcRbZ1OvxXXZvz5BYbTl3PE56XzTLIi8IARIrLp7UwwGah7wRIPlqwjTGv5gETemvBOHA0u81/qdJN2/hponC/hSVNrreICIxCAESCjCEo5UCmoe8ESAaISApXlWhGx7YJ0W7msFB8UZ3ccb34A2wituHSDNNwC+vUiIvCAESKzKUr4kHmoe8ESBRitBAQYRUwIuwxToWGUuykjSQ/3v+UTukcYp8GLfDxCAiMQgBEgo2+rCGC5qHvBEgGiEgLj/3F/ldj3NVCt093bu3KkvnZmIdyiUFebMuvuyUbzIiMQgBEgo6pO7DF5qHvBEgGiEg9zM5uIHVIuir4EP+gIsgbbBOVmzFGA7x/584qrRfCWMiLwgBEis82KjiKZqHvBEgv+vFEu8mfySMTTMpCxrAsa29dWdr/XENEhIHR0IjQxIgCqcCCqQCCgNpYmMSIKQ3d82RhIswOYN31gMNoMchRBY5NefMSA/gy/EsSRwYGgkIARgBIAEqAQAiJQgBEiEBKjuiiNqJvlljx71xY2yq/iwVFVQxPrOSvfQBKgnDV2UiJwgBEgEBGiBHXvjauGRi45rrKP1iWQShz3IwsOMG1Z5+iPG4nkcMlCInCAESAQEaIPMtbm+gZvToJzoTgAKaxI5C4uzbvnRGDJSDLuzj+qoRIicIARIBARogqQl+dp6NhYXg8dzt5fh9GUxhdUYuvuuJdFPSSgCeyxoiJQgBEiEBTlU3ks5ni5QvLzArwPjJBc57fSvHKDnoIkxqGelBGf0iJwgBEgEBGiCiLJQhelNBQRLbD2WVDmW7gw7EhmAEDi3qR3Q9RmjkEw==");
```
Comment (robert-zaremba) [inline src/context.rs]: how is this generated? Could you provide a comment above this variable?

Comment (vuvoth) [reply src/context.rs]: Yes! This from [mintscan tx](https://www.mintscan.io/cosmos/tx/A0E69441FB46C5797C1193D6EAA7EB5A59A809F0433ECA6CE29D7CD3DEFED679?height=21413592&sector=json).

Comment (vuvoth) [reply src/context.rs]: I updated comment for this test

File: README.md (hunk 1 above)
Comment (vuvoth) [reply README.md]: I separate verifying and updating command. I also added format about Consensus State in README.

File: README.md [hunk 4]
```diff
@@ -1,3 +1,37 @@
 # Tendermint Light Client
 
 PoC for a minimalistic TM Light Client based on [lib-rs](https://github.com/cosmos/ibc-rs)
+
+## Cli interface 
+
+### Fetch Consensus State and Header 
+This command fetch consensus state and header from full-node. This only use for testing `Verify` command.
```
Comment (robert-zaremba) [inline README.md:9]:
```suggestion

This command fetch consensus state and header from full-node. This only use for testing `Verify` command.  
```

File: README.md [hunk 5]
```diff
@@ -1,3 +1,37 @@
 # Tendermint Light Client
 
 PoC for a minimalistic TM Light Client based on [lib-rs](https://github.com/cosmos/ibc-rs)
+
+## Cli interface 
+
+### Fetch Consensus State and Header 
+This command fetch consensus state and header from full-node. This only use for testing `Verify` command.  
+
```
Comment (robert-zaremba) [inline README.md:10]: command is missing here

File: README.md [hunk 6]
```diff
@@ -1,3 +1,37 @@
...
+```rust
+pub struct ConsensusState {
+    pub timestamp: Time,
+    pub root: CommitmentRoot,
+    pub next_validators_hash: Hash,
+}
+```
+
+This command verifies a new state (can extract from header) is valid state start from consensus state in cs_path file.
```
Comment (robert-zaremba) [inline README.md]:
```suggestion

This command verifies a new state (can extract from header) is valid state start from consensus state in cs_path file.

```bash
tendermint-lightclient verify <CS_PATH> <HEADER_PATH>
```

The Consensus State format follow [ics-007-tendermint-client](https://github.com/cosmos/ibc/tree/main/spec/client/ics-007-tendermint-client#consensus-state) and `root` is `app_hash`. Other data can easy to get from block header.


```rust
pub struct ConsensusState {
    pub timestamp: Time,
    pub root: CommitmentRoot,
    pub next_validators_hash: Hash,
}
```
```

File: README.md [hunk 7]
```diff
@@ -1,3 +1,37 @@
...
+
+### Update 
+```bash
+tendermint-lightclient update <CS_PATH> <HEADER_PATH> <NEW_CS_PATH>
+```
+This command work like verify command but create new consensus state and save to new_cs_path.
```
Comment (robert-zaremba) [inline README.md]:
```suggestion

This command work like verify command but create new consensus state and save to new_cs_path. 

```bash
tendermint-lightclient update <CS_PATH> <HEADER_PATH> <NEW_CS_PATH>
```
```

File: src/context.rs [hunk 8]
```diff
@@ -294,4 +301,75 @@ mod tests {
...
+            params.allow_update,
+        )
+        .unwrap();
+
+        let client = ClientState::from(client);
+
+        // Data for this test from this tx:
+        //  - https://www.mintscan.io/cosmos/tx/A0E69441FB46C5797C1193D6EAA7EB5A59A809F0433ECA6CE29D7CD3DEFED679?height=21413592&sector=json
+        // This transaction transfer token from Osmosis to Cosmos Hub
+        // ibc module prefix ons
```
Comment (robert-zaremba) [inline src/context.rs]:
```suggestion
        // ibc module prefix = "ibc"
```

File: src/context.rs [hunk 9]
```diff
@@ -294,4 +301,75 @@ mod tests {
...
...N5yIwCAESCSCu0wKah7wRIBohIEpnFv1GjFbFsqE429zfZiKfOT6//q45DUFSPLrY9QoDIi4IARIqIpjxBZqHvBEgQaWrGBvLw429u+E3zsCC20seuQ2K4Np8CgTQd8VhLEQgIjAIARIJJKTOB5qHvBEgGiEg/7viCbl3Q1jvaUZEPFz/2gw7knpIfR+pUpQfbqemvM8iLggBEiomuKMPmoe8ESBx0HzWr2/H5w2JBV7mGKRoCT2nY3baDGbXYLiB/hIYoCAiMAgBEgkq6Jwhmoe8ESAaISAgRLXnK55+pWAZPGrk6b1ZPtOwMcyynsQFqaJmY1SkyiIwCAESCSygyUmah7wRIBohIDzvjMPa1Sr0H1VwcRbZ1OvxXXZvz5BYbTl3PE56XzTLIi8IARIrLp7UwwGah7wRIPlqwjTGv5gETemvBOHA0u81/qdJN2/hponC/hSVNrreICIxCAESCjCEo5UCmoe8ESAaISApXlWhGx7YJ0W7msFB8UZ3ccb34A2wituHSDNNwC+vUiIvCAESKzKUr4kHmoe8ESBRitBAQYRUwIuwxToWGUuykjSQ/3v+UTukcYp8GLfDxCAiMQgBEgo2+rCGC5qHvBEgGiEgLj/3F/ldj3NVCt093bu3KkvnZmIdyiUFebMuvuyUbzIiMQgBEgo6pO7DF5qHvBEgGiEg9zM5uIHVIuir4EP+gIsgbbBOVmzFGA7x/584qrRfCWMiLwgBEis82KjiKZqHvBEgv+vFEu8mfySMTTMpCxrAsa29dWdr/XENEhIHR0IjQxIgCqcCCqQCCgNpYmMSIKQ3d82RhIswOYN31gMNoMchRBY5NefMSA/gy/EsSRwYGgkIARgBIAEqAQAiJQgBEiEBKjuiiNqJvlljx71xY2yq/iwVFVQxPrOSvfQBKgnDV2UiJwgBEgEBGiBHXvjauGRi45rrKP1iWQShz3IwsOMG1Z5+iPG4nkcMlCInCAESAQEaIPMtbm+gZvToJzoTgAKaxI5C4uzbvnRGDJSDLuzj+qoRIicIARIBARogqQl+dp6NhYXg8dzt5fh9GUxhdUYuvuuJdFPSSgCeyxoiJQgBEiEBTlU3ks5ni5QvLzArwPjJBc57fSvHKDnoIkxqGelBGf0iJwgBEgEBGiCiLJQhelNBQRLbD2WVDmW7gw7EhmAEDi3qR3Q9RmjkEw==");
```
Comment (robert-zaremba) [inline src/context.rs]: we already have this data in `src/data/proof` file. let's load it from there, rather copying it here.

File: src/context.rs [hunk 10]
```diff
@@ -294,4 +301,75 @@ mod tests {
...
+        // This is the root of multistore or app_hash/root of Osmosis client on CosmosHub
+        let root = CommitmentRoot::from_bytes(&base64_to_bytes(
+            "QNvnS2A//laHEjW9qVB5hya4zOCyugOIXU4vi0vnBww=",
+        ));
+        // Those data help us get the path of Commitment Path. You can check packet field in MsgRecvPacket msg.
+        let port_id = PortId::new("transfer".to_owned()).unwrap();
+        let channel_id = ChannelId::new(0);
+        let sequence = Sequence::from(3514632);
+
+        // data, timeout_height and timeout_timestamp from packet field in MsgRecvPacket msg.
```
Comment (robert-zaremba) [inline src/context.rs]:
```suggestion
        // IBC MsgRecvPacket type fields:
```

File: src/main.rs [hunk 11]
```diff
@@ -1,10 +1,161 @@
...
+use utils::{base64_to_bytes, fetch_consensus_state, fetch_header, CSReadable};
+
 mod api;
 mod context;
+mod provider;
 mod storage;
+mod utils;
+
+#[derive(Parser, Debug)]
+enum LCCLi {
```
Comment (robert-zaremba) [inline src/main.rs]: what `LCCLi` stands for? Could you add a comment or find better name (eg `CLIcommand`)?

**Review by robert-zaremba — APPROVED** [review]

File: src/main.rs (hunk 11 above)
Comment (vuvoth) [reply src/main.rs]: A LCCli stands for Light Client Command Line.

File: README.md (hunk 5 above)
Comment (vuvoth) [reply README.md:10]: I updated it.

File: src/context.rs (hunk 9 above)
Comment (vuvoth) [reply src/context.rs]: Done

**Pull request merged**